File: Linked List/circular_link_list/js/circularLL.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class circularLL {
  constructor() {
    this.head = null;
    this.count = 0;
  }

  //METHOD TO INSERT AT BEGINNING
  insert_at_beginning(data) {
    let new_node = new Node(data);
    // new_node.next = new_node;

    this.count++;
    if (this.count == 1) {
      this.head = new_node;
      this.head.next = this.head;
      return;
    }

    let ptr = this.head;
    while (ptr.next != this.head) ptr = ptr.next;

    new_node.next = ptr.next;
    ptr.next = new_node;
    this.head = new_node;
  }

  //METHOD TO INSERT AT END
  insert_at_end(data) {
    let new_node = new Node(data);
    new_node.next = new_node;

    this.count++;
    if (this.count == 1) {
      this.head = new_node;
      return;
    }

    let ptr = this.head;
    while (ptr.next != this.head) ptr = ptr.next;

    new_node.next = ptr.next;
    ptr.next = new_node;
  }

  //METHOD TO INSERT AT ANY POSITION
  insert_at_position(position, data) {
    if (position < 1 || position > this.count + 1) {
      console.log("Invalid Position");
      return;
    }

    if (position == 1) {
      this.insert_at_beginning(data);
      return;
    }

    if (position == this.count + 1) {
      this.insert_at_end(data);
      return;
    }

    this.count++;

    let new_node = new Node(data);
    let ptr = this.head;
    let index = 1;
    position--;

    while (index != position) {
      index++;
      ptr = ptr.next;
    }

    new_node.next = ptr.next;
    ptr.next = new_node;
  }

  //METHOD TO DELETE FROM BEGINNING
  delete_from_beginning() {
    if (this.count == 0) {
      console.log("Empty list");
      return;
    }

    this.count--;
    if (this.count == 0) {
      this.head = null;
      return;
    }

    let ptr = this.head;
    let temp = this.head;
    while (ptr.next != this.head) ptr = ptr.next;

    ptr.next = this.head.next;
    this.head = this.head.next;
    temp.next = temp;
  }

  //METHOD TO DELETE FROM END
  delete_from_end() {
    if (this.count == 0) {
      console.log("Empty list");
      return;
    }

    this.count--;
    if (this.count == 0) {
      this.head = null;
      return;
    }

    let ptr = this.head;

    while (ptr.next.next != this.head) ptr = ptr.next;

    let temp = ptr.next;
    ptr.next = this.head;
    temp.next = temp;
  }

  //METHOD TO DELETE FROM A POSITION
  delete_from_position(position) {
    if (position < 1 || position > this.count) {
      console.log("Invalid Position");
      return;
    }

    if (this.count == 0) {
      console.log("Empty list");
      return;
    }

    if (position == 1) {
      this.delete_from_beginning();
      return;
    }

    if (position == this.count) {
      this.delete_from_end();
      return;
    }

    this.count--;
    let ptr = this.head;
    let index = 1;
    position--;
    while (index != position) {
      index++;
      ptr = ptr.next;
    }
    let temp = ptr.next;
    ptr.next = temp.next;
    temp.next = temp;
  }

  //METHOD TO PRINT THE LIST
  print_list() {
    if (this.count == 0) {
      console.log("empty List.");
      return;
    }

    let ptr = this.head;
    process.stdout.write(ptr.data + " -> ");
    while (ptr.next != this.head) {
      ptr = ptr.next;
      process.stdout.write(ptr.data + " -> ");
    }
  }
}

module.exports = {
  circularLL,
};

File: Data structure/Linked List/circular_link_list/python/circularLL.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class circularLL:
    def __init__(self):
        self.head = None
        self.count = 0

    # METHOD TO INSERT AT BEGINNING
    def insert_at_beginning(self, data):
        new_node = Node(data)
        new_node.next = new_node

        self.count += 1

        if(self.head == None):
            self.head = new_node
            return

        ptr = self.head
        while(ptr.next != self.head):
            ptr = ptr.next

        new_node.next = self.head
        ptr.next = new_node
        self.head = new_node

    # METHOF TO INSERT AT END
    def insert_at_end(self, data):
        new_node = Node(data)
        new_node.next = new_node

        self.count += 1

        if(self.head == None):
            self.head = new_node
            return

        ptr = self.head
        while(ptr.next != self.head):
            ptr = ptr.next

        new_node.next = self.head
        ptr.next = new_node

    # METHOD TO INSERT AT GIVEN POSITION
    def insert_at_position(self, position, data):
        if(position < 1 or position > self.count+1):
            print("Invalid position.")
            return

        if(position == 1):
            self.insert_at_beginning(data)
            return

        if(position == self.count+1):
            self.insert_at_end(data)
            return

        self.count += 1
        index = 1
        position -= 1
        new_node = Node(data)

        while(index != position):
            index += 1
            ptr = ptr.next

        new_node.next = ptr.next
        ptr.next = new_node

    # METHOD TO DELETE FROM BEGINNING
    def delete_from_beginning(self):
        if(self.count == 0):
            print("empty list.")
            return

        self.count -= 1

        if(self.count == 0):
            self.head = None
            return

        ptr = self.head
        while(ptr.next != self.head):
            ptr = ptr.next

        ptr.next = self.head.next
        self.head = ptr.next

    # METHOD TO DELET FROM END
    def delete_from_end(self):
        if(self.count == 0):
            print("empty list.")
            return

        self.count -= 1

        if(self.count == 0):
            self.head = None
            return

        ptr = self.head
        while(ptr.next.next != self.head):
            ptr = ptr.next

        ptr.next.next = None
        ptr.next = self.head

    # METHOD TO DELETE FROM A GIVEN POSITION
    def delete_from_position(self, position):
        if(position < 1 or position > self.count):
            print("Invalid position.")
            return

        if(position == 1):
            self.delete_from_beginning()
            return

        if(position == self.count):
            self.delete_from_end()
            return

        self.count -= 1
        index = 1
        position -= 1
        ptr = self.head
        while(index != position):
            index += 1
            ptr = ptr.next

        ptr.next = ptr.next.next

    # METHOD TO PRINT THE LIST
    def print_list(self):
        if self.count == 0:
            print("Empty list")
            return

        ptr = self.head
        print(ptr.data, " -> ", end="")

        while(ptr.next != self.head):
            ptr = ptr.next
            print(ptr.data, " -> ", end="")
        print()
